### PR TITLE
Fix version comparisons for 4-segment pins like opencv-python!=4.13.0.90

### PR DESCRIPTION
**Pull request by glenn-jocher**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Improves version parsing and comparison reliability, especially for 4-part dependency versions and Jetson TensorRT exports 🚀

### 📊 Key Changes
- Updated `parse_version()` to preserve all numeric release segments, such as `4.13.0.92`, instead of truncating to 3 parts 🔢
- Improved `check_version()` to compare versions of different lengths accurately by padding them before comparison ✅
- Expanded test coverage for versions with prefixes, suffixes, pre-release tags, local build tags, and 4-segment pins 🧪
- Adjusted TensorRT export logic for JetPack 6 to match the full TensorRT `10.3.x` family instead of only `10.3.0` ⚙️
- Updated system info collection to avoid raising hard errors during dependency checks, making diagnostics more robust 🛠️

### 🎯 Purpose & Impact
- Prevents incorrect dependency checks when packages use 4-part version numbers, improving compatibility and reliability 📦
- Fixes Jetson TensorRT quantized export handling for JetPack 6 builds that report versions like `10.3.0.x` 🤖
- Reduces false positives or false negatives in version constraints such as exact matches, exclusions, and range checks 🎯
- Makes system diagnostics more user-friendly by reporting unmet requirements without interrupting execution 🧩

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message is here to help ensure a seamless integration of your dependency version comparison fix; an Ultralytics engineer will also assist with review.

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/utils/checks.py [hunk 1]
```diff
@@ -116,14 +116,14 @@ def parse_version(version="0.0.0") -> tuple:
...
+        version (str): Version string, i.e. '2.0.1+cpu' or '4.13.0.92'
 
     Returns:
-        (tuple): Tuple of integers representing the numeric part of the version, i.e. (2, 0, 1)
+        (tuple): Tuple of integers representing the numeric part of the version, at least 3 long, i.e. (2, 0, 1)
     """
     try:
-        nums = [int(x) for x in re.findall(r"\d+", version)[:3]]
-        return tuple(nums + [0] * (3 - len(nums)))  # pad to 3, i.e. '2.0.1+cpu' -> (2, 0, 1), '2' -> (2, 0, 0)
+        nums = [int(x) for x in re.findall(r"\d+", version)]
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/checks.py]: 💡 **MEDIUM**: Collecting every digit group now includes numeric local/build suffixes such as PyTorch's `2.0.1+cu118`. That changes `parse_version('2.0.1+cu118')` to `(2, 0, 1, 118)`, so checks like `check_version(torch.__version__, '==2.0.1')` or `<=2.0.1` can fail even though the public version is `2.0.1`. The fix should preserve dot-separated release segments like `4.13.0.90` while still ignoring non-release suffixes.

**Suggested change:**
```suggestion
        nums = [int(x) for x in re.match(r"\d+(?:\.\d+)*", version).group(0).split(".")]
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The tuple-padding comparison change addresses the 4-segment pin issue, and the system-info hard-failure adjustment looks reasonable. One regression remains in `parse_version()`: it now treats numeric local/build suffixes as release segments, which can break equality and upper-bound checks for common versions like `+cu118`.

💬 Posted 1 inline comment

File: ultralytics/utils/checks.py (hunk 1 above)
Comment (glenn-jocher) [reply ultralytics/utils/checks.py]: Fixed in 7b25abf65 — `parse_version` now matches only the leading dot-separated release segments (`re.match(r"\d+(?:\.\d+)*", ...)`), so `2.0.1+cu118` → `(2, 0, 1)` and `1.0.0rc1` → `(1, 0, 0)` while `4.13.0.92` keeps all 4 segments. Added test assertions for both suffix cases.

**Comment by UltralyticsAssistant** [comment]
🚀 Merged — thank you @glenn-jocher!

“Details make perfection, and perfection is not a detail.” — Leonardo da Vinci

This PR is a great example of that principle in action: improving version parsing, strengthening dependency comparisons, expanding edge-case test coverage, and making Jetson TensorRT export handling more reliable for JetPack 6. These refinements help reduce false positives, improve diagnostics, and make Ultralytics workflows more robust across diverse environments.

Appreciate the careful attention to compatibility and reliability! 🙌
